x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/pki/X509AuthenticationToken.java

```diff
@@ -15,21 +16,21 @@
 
     private final String dn;
     private final X509Certificate[] credentials;
-    private final boolean isDelegated;
+    private final Authentication delegateeAuthentication;
     private String principal;
 
     public X509AuthenticationToken(X509Certificate[] certificates) {
-        this(certificates, false);
+        this(certificates, null);
     }
 
-    public X509AuthenticationToken(X509Certificate[] certificates, boolean isDelegated) {
+    public X509AuthenticationToken(X509Certificate[] certificates, Authentication delegateeAuthentication) {
```

Comment:
WDYT of making this constructor private, and adding a factory method like:
```
public static X509AuthenticationToken delegatedToken(X509Certificate[] certificates, Authentication delegateeAuthentication) {
  Objects.requireNotNull(delegateeAuthentication);
 return new X509AuthenticationToken(certificates, delegateeAuthentication);
}
``` 

I worry about constructors where passing in a `null` is allowed, but completely changes the behaviour. You can't guard against null (because it's allowed) but if someone accidentally passes in a null value then they will get a behaviour that doesn't do what they want (and could end up being a security issue).

Reply:
Great idea to make creating delegated tokens more explicit.